common/processes/platform/tiny_process_library/tiny_process_library.cpp

```diff
@@ -77,23 +215,89 @@ ExpectedLineData Process::GenerateLineData() {
 			CollectLineData(trailing_line, ret, bytes, len);
 		});
 
-	auto id = proc_->get_id();
+	if (proc_->get_id() == -1) {
+		proc_.reset();
+		return expected::unexpected(MakeError(
+			ProcessesErrorCode::SpawnError,
+			"Failed to spawn '" + (args_.size() >= 1 ? args_[0] : "<null>") + "'"));
+	}
 
-	// waits for the process to finish
-	// TODO: log exit status != 0? Or error?
-	Wait();
+	SetupAsyncWait();
+
+	auto status = Wait(timeout);
+	if (!status) {
+		return expected::unexpected(
+			MakeError(NonZeroExitStatusError, "Collecting line data failed"));
+	}
 
 	if (trailing_line != "") {
 		ret.push_back(trailing_line);
 	}
 
-	if (id == -1) {
-		return expected::unexpected(MakeError(
-			ProcessesErrorCode::SpawnError,
-			"Failed to spawn '" + (args_.size() >= 1 ? args_[0] : "<null>") + "'"));
+	return ExpectedLineData(ret);
+}
+
+io::ExpectedAsyncReaderPtr Process::GetProcessReader(events::EventLoop &loop, int &pipe_ref) {
+	if (proc_) {
+		return expected::unexpected(
+			MakeError(ProcessAlreadyStartedError, "Cannot get process output"));
 	}
 
-	return ExpectedLineData(ret);
+	if (pipe_ref >= 0) {
+		close(pipe_ref);
+		pipe_ref = -1;
+	}
+
+	int fds[2];
+	int ret = pipe(fds);
+	if (ret < 0) {
+		int err = errno;
+		return expected::unexpected(error::Error(
+			generic_category().default_error_condition(err),
+			"Could not get process stdout reader"));
+	}
+
+	pipe_ref = fds[1];
+
+	return make_shared<events::io::AsyncFileDescriptorReader>(loop, fds[0]);
+}
+
+io::ExpectedAsyncReaderPtr Process::GetAsyncStdoutReader(events::EventLoop &loop) {
+	return GetProcessReader(loop, stdout_pipe_);
+}
+
+io::ExpectedAsyncReaderPtr Process::GetAsyncStderrReader(events::EventLoop &loop) {
+	return GetProcessReader(loop, stderr_pipe_);
+}
+
+int Process::EnsureTerminated() {
+	if (proc_) {
+		log::Info("Sending SIGTERM to PID " + to_string(proc_->get_id()));
```

Comment:
Q: Should this really be as high as info?

Reply:
I think in most cases we expect processes to return by themselves, so logging that we are sending signals is worthwhile, I think.